[FIX] TMTTenPlexQuantiationMethod -  mass mapping

oliveralka opened the pull request:
It seem that there are at least two different correction maps for TMT10Plex (Thermo). 
Here a choice gets introduced in the TMTThenPlexQuantiationMethod.

Here: correction map - relationship between the reporterions

correction map 1: [1]https://tools.thermofisher.com/content/sfs/COAPDFs/2015/OJ185435_90110.pdf
correction map 2: [2]https://tools.thermofisher.com/content/sfs/COAPDFs/2015/SPECS_90111.pdf

cbielow commented:
this seems to collide heavily with the correction_matrix approach.
Why would you hard-code these values?

oliveralka commented:
The correction map was hard-coded in the first place. I only added a second outline of the map. The corrections matrix (percentage) is still needed to gain the percentage of the reporterion-relation. I am in contact with Thermo, if there are more maps than the two - I can still soft-code it like the correction matrix - then the user has to fill out more forms - it is - in my opinion - more practical this way.

cbielow commented:
you added both hard-coded versions here. Where is the 'first' one?

Also, how would the user use the soft-coded one? It seems that 'correction_matrix' is ignored completely in your approach, which must at least be documented to the user.
In case there is a third alternative in the future, and OpenMS does not have it yet -- this would also mean a) constant checking on our side for new versions and 
b) old OpenMS releases will not support this, i.e. you need to provide a fallback mechanism

If you want to implement sensible defaults, thats fine. But make sure that the user can change them, in case the defaults become outdated.

oliveralka commented:
This may be a misunderstanding. If I understood the code correctly, there are two kinds of matrix. 

First the "correction map" with the name _TMTTenPlexQuantiationMethod_. Here is shown which repoterion is mapped to another, dependent on the mass difference. (For the original map see: https://github.com/OpenMS/OpenMS/blob/develop/src/openms/source/ANALYSIS/QUANTITATION/TMTTenPlexQuantitationMethod.cpp).  The first map is not "original" since I tweaked the masses of the reporter ions as described in [1] & [2] (126 - 126.127725 -> 126 - 126.127726). Here the user can switch between two different map layouts (correction_map_1, correction_map_2). Correction_map_1 is the default value, which conforms the original one (with a slight alteration of the masses).

The second matrix is the correction matrix, where the user has to specify at which percentage the repoterions differ.

So the user has to choose between correction_map_1 (layout_1) and correciont_map_2 (layout_2) (see IsobaricAnalyzer in this fix) and has to specify the correction matrix with the percentage indicated as shown in their product data sheet. 

Of course it may be that I made an implementation error or have a wrong understanding of the code.

hendrikweisser commented:
> If I understood the code correctly, there are two kinds of matrix.

Without having looked at the code: If your understanding is correct, then could you rename the "correction map" (which contains the accurate masses) to something else (e.g. "mass mapping")? There doesn't really seem to be any "correction" involved in the use of this map, so the name is misleading - and confusing because of the clash with the (more aptly named) "correction matrix".

hroest commented:
I clicked on the two links and in the second link I only see all values replaced by red asterisks - so maybe this was replaced recently? Or the second one is just a template? Anyways, I dont really see two different matrices here. 

So first, I would like to understand why there are 2 matrices, there seems to be one where all C only interact with C plexes and all N only with N (e.g. 128C only interacts with 126, 127C, 129C, 130C) which makes sense if the mass accuracy is sufficient -- I guess? But this datasheet (90111) seems to be actually empty only with red asterisks in it. However, why shouldnt we have a more generic approach in which all interactions are specified and then set for zero for those that dont interact so that the user can select how they want to set the interactions? Of course there should be a sensible default based on the current datasheet.

I am still a bit confused. Did Thermo change the composition of the product between 90110 and 90111 or did they just change the calculation method? If the latter is the case, wouldnt it make more sense to empirically determine which matrix works better / is more accurate? 

Btw: just to improve my understanding, this is to remove the effect that the reporter ions are not 100% pure, right?

oliveralka commented:
The second link was a specs sheet (template) to show that is only about the mapping of the masses. In my opinion all new charges have this kind of mapping [3]. 

I don't know why they have to different mappings. The second mapping [2 or 3] makes more sense in my opinion. 

You suggest one matrix for the mapping and a second matrix for the correction, which can be filled by the user? If there are more than two mappings or if this mapping changes in the future than this would make sense. If it stays with these two mapping, than just hard-code them would be easier. 

I am waiting for a reply from Thermo. 

> Btw: just to improve my understanding, this is to remove the effect that the reporter ions are not 100% pure, right?

As far as I understood - yes. 

[3] https://tools.thermofisher.com/content/sfs/COAPDFs/2015/PB195275A_90406.pdf

cbielow commented:
there was indeed some confusion, mostly on my part :) sorry for that. After looking at the two mappings from Thermo, it seems they differ only at very few positions (not talking about the isotope contamination values-intensities, but the mapping). The first one (https://tools.thermofisher.com/content/sfs/COAPDFs/2015/OJ185435_90110.pdf) seems just wrong, since it has `0 (128C)` as 2nd isotope of 127N (which clearly is only 1 Da apart). There are 2-3 similar glitches in there. So the question to Thermo then is: is only the naming wrong and the values are correct? (this would mean we only need one mapping table).

concerning the new correction mappings and exact masses:
It should go into a separate function (for both mappings) and be called where appropriate.
This also disambiguates the mass accuracy isues (`126.12772(5|6)`) , even though this does not make any practical difference on any modern mass spec (it's 0.0077ppm)

cbielow commented:
according to
https://tools.thermofisher.com/content/sfs/gallery/high/90110-NHS-TMT10-stc.jpg

heavy C and N can indeed be mixed, i.e. depending on the synthetization approach, imputities of all "N" channels can affect all other channels. Thus, the simple -2, -1, 0, +1, +2 scheme might not even be enough.
The better way to model this would be to interpret the matrix as channels ordered by exact mass, instead of just using the nominal mass (which is ambiguous for TMT10plex, as opposed to iTRAQ). This will require more columns since within 2 Da, there are up to 5 channels in each direction. Or specify the complete matrix (10x10) matrix.

hroest commented:
actually, this is what I was describing: assume a full matrix and simply set the values that are not important to zero (optimally user provided). This seems to be a much more sensible approach than to hardcode the matrix which may or may not be accurate. By setting the appropriate values to zero, the user can simply generate whatever interactions he/she wants ...

cbielow commented:
there is downsides to changing the matrix layout though: the user cannot simply copy and paste the tmt-certificate he gets, but has to make the translation manually (and get it right).

oliveralka commented:
Thermo answered: " Die ursprüngliche Datei, der Vorlage ohne eingetragene Werte [2], weist einen Fehler auf. Dieser Fehler in der CoA Vorlage, wurde Anfang 2016 behoben."

That means that the specs sheet before 2016 had an error. As far as I understood, [2] is the corrected version. Therefore we have differences between [1] and [2].

In the end we need only one mass mapping for the TMTTenPlexQuantitationMethod as before. The only thing is that the mapping has to be adjusted as shown in mass maps outline (2).

```
       //    // mass map outline (2)
        //    "126", 126.127726, x, x, 127C, 128C
        //    "127N", 127.124761, x, x, 128N, 129N
        //    "127C", 127.131081, x, 126, 128C, 129C
        //    "128N", 128.128116, x, 127N, 129N, 130N
        //    "128C", 128.134436, 126, 127C, 129C, 130C
        //    "129N", 129.131471, 127N, 128N, 130N, 131
        //    "129C", 129.137790, 127C, 128C, 130C, x
        //    "130N", 130.134825, 128N, 129N, 131, x
        //    "130C", 130.141145, 128C, 129C, x, x
        //    "131", 131.138180, 129N, 130N, x, x
```

oliveralka commented:
I think this can be seen as fix. Of course it can be further discussed if this should be soft coded in the near future. In my opinion this would make sense, if the TMT10Plex mapping is to be changed, which I do not assume.

hroest commented on src/openms/source/ANALYSIS/QUANTITATION/TMTTenPlexQuantitationMethod.cpp:
```diff
@@ -45,37 +45,49 @@ using namespace boost::assign;
+{
     setName("TMTTenPlexQuantitationMethod");
 
+    //    // mass map outline (2)
```
maybe add a comment linking to this thread and dont call it `(2)`

cbielow commented:
ok, looks good to me and glad we can have the user-friendly solution, i.e. no additional parameter required.

Just one thing: what about tests? Should this change not trigger a failed test? Could you add one, which shows that the mapping done here is actually working as expected, i.e. how do we know the channels are mapped correctly?

oliveralka commented:
@cbielow I have changed the test from the previous mapping - if that was what you are looking for - please see commit 9e49ed7.

cbielow commented:
I saw that you check that the mapping is correct. However, this does not mean the actual data is handled correctly, right? So plugging in an actual MS2 spectrum with data was what I was asking for :)

oliveralka commented:
Ok, that makes sense. I will provide one for the TMT10Plex data.

oliveralka commented:
`Continuous build .4: The job exceeded the maximum time limit for jobs, and has been terminated.` 

Is the time limit too stringent or what may be the issue?

jpfeuffer commented:
Yep. Fails regularly with Clang since some months. Usually nothing to worry but you can restart the check once to give it another chance.

hroest commented on src/openms/source/ANALYSIS/QUANTITATION/TMTTenPlexQuantitationMethod.cpp:
```diff
@@ -90,27 +102,23 @@ namespace OpenMS
+                                                                      "0.0/0.0/0.0/0.0,"
+                                                                      "0.0/0.0/0.0/0.0,"
+                                                                      "0.0/0.0/0.0/0.0"),
                        "Correction matrix for isotope distributions (see documentation); use the following format: <-2Da>/<-1Da>/<+1Da>/<+2Da>; e.g. '0/0.3/4/0', '0.1/0.3/3/0.2'");
```
maybe unrelated but the whole "see documentation" is not so great : where actually is the documentation? I would expect this to be the tool docu http://ftp.mi.fu-berlin.de/pub/OpenMS/develop-documentation/html/TOPP_IsobaricAnalyzer.html but there is not much more information there (except "see documentation" again!). So if I am new to OpenMS, how would I ever figure out what values to put there? E.g. it is clear that its -2,-1,+1, +2 but its never specified which channels these actually are -- except in the source code?

cbielow replied:
this is the scheme Thermo uses in their certificates, so it should be clear to the user what to fill in.
But I agree, it could be more explcitily stated in the docs.
Maybe a doc-fix hour during the retreat would be good...

cbielow commented:
ahh ok. So you're testing on TOPP level. Fine IMHO if you know the results are fine. Usually its more intuitive to do this on actual spectral level (here ` void IsobaricChannelExtractor::extractChannels(..)` should do the trick), but either works I guess.

oliveralka commented:
@cbielow I have another issue (if needed I can open another one). 
The Isobaric Analyzer fails for some files. It happens while computing the  Stats and fails with the Error: 

> IsobaricIsotopeCorrector: Isotope correction values of alternative method differ!

Implemented was the absolute difference: 
```
      // Absolute
      // else if (std::fabs(m_x(index, 0) - x(index)) > 0.000001)
```
Since the absolute difference in the intensity was bigger for example:
fabs(24096 - 24153) =  57
75 > 0.000001

I replaced it by the relative difference, which should not be greater than 1%
This would allow a higher variance the higher the intensity.
```
      // Relative  1%
      else if ((((std::fabs(m_x(index, 0) - x(index)))/m_x(index, 0))*100) > 1)
```
fabs(24096 - 24153) =  57
(75/24096)*100 = 0,311%
0,311% > 1%

I wanted to ask if this is a sufficient solution or why it should be > 0.000001?
I am not sure which variances between the inverse matrix and the NNLS model can be allowed. 

Original Method: 
```
  IsobaricIsotopeCorrector::computeStats_(const Matrix<double>& m_x,
                                          const Eigen::MatrixXd& x, const float cf_intensity,
                                          const IsobaricQuantitationMethod* quant_method, IsobaricQuantifierStatistics& stats)
  {
    Size s_negative(0);
    Size s_different_count(0); // happens when naive solution is negative in other channels
    double s_different_intensity(0);

    // ISOTOPE CORRECTION: compare solutions of Matrix inversion vs. NNLS
    for (Size index = 0; index < quant_method->getNumberOfChannels(); ++index)
    {
      if (x(index) < 0.0)
      {
        ++s_negative;
      }
      else if (std::fabs(m_x(index, 0) - x(index)) > 0.000001)
      {
        ++s_different_count;
        s_different_intensity += std::fabs(m_x(index, 0) - x(index));
      }
    } 
```
```
 if (s_negative == 0 && s_different_count > 0) // solutions are inconsistent, despite being positive! This should not happen!
    {
      throw Exception::Postcondition(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "IsobaricIsotopeCorrector: Isotope correction values of alternative method differ!");
    }
```

cbielow commented:
the error message you are seeing appears only if both solutions should be identical, but are in fact not.
In theory, NNLS delivers the exact same result as Eigen if the solution is entirely positive.
The check is there to simply verify that and it should never throw if both methods were stable.
Apparently, this is not the case with your data. So there is either a severe bug in either NNLS or Eigen, or its simply numeric instability.
To test which one is correct, it should be possible to multiply the result matrix with the correction matrix and get back the observed data.

timosachsenberg commented:
In MetaProSIP that situation arose quite frequently that an LS method produced other results as the NNLS method. The reason was exclusively what @cbielow suggested as a possible cause: the LS produced negative coefficients (and at the same time larger positive coefficients than NNLS). A direct effect is that NNLS produces sparser solutions (less wobbling between negative and positive coefficients) in the presence of noise. There are also some theoretical results that support this observation.
On the other hand, the LS solution has a smaller squared error if one would allow negative coefficients.
My take on this is that the NNLS solution should be chosen in the case both deviate (which actually means that only NNLS is used).

cbielow commented:
well, thats only somewhat what I meant: if LS produces negative results, then NNLS is bound to differ (since it cannot be negative by definition). But if NS produces positive results, NNLS should be identical. Exactly this is not the case in practice here and needs some investigation (stability problem?).

The rationale behind usage of NNLS for reporter ions, is that it is impossible to have negative amounts of a substance, hence we restrict the solution to positive values. The reasons for this can be manifold (mostly ion statistics of the detector (saturation, lack of sensitivity etc) or other contaminations -- at least for iTRAQ they exist).

timosachsenberg commented:
Ah ok sorry you are right - should have read more carefully.

timosachsenberg commented:
I would suggest that we accept the fix that compares by relative deviation to get it working with Oliver's data. Might be worth digging into this problem if it reoccurs, though. Spending too much time on a problem that only showed up in a neglectable amount of spectra sounds like overkill at that point.

cbielow commented:
sure, fine with me

oliveralka commented:
Will test again and commit asap.

timosachsenberg commented:
+1
